cmd/k8s-operator/svc-for-pg.go

```diff
@@ -0,0 +1,899 @@
+			podIPv4 = parsed.String()
+			continue
+		}
+		podIPv6 = parsed.String()
+	}
+	if podIPv4 != gotCfgs.PodIPv4 || podIPv6 != gotCfgs.PodIPv6 {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (a *HAServiceReconciler) maybeUpdateAdvertiseServicesConfig(ctx context.Context, svc *corev1.Service, pgName string, serviceName tailcfg.ServiceName, cfg *ingressservices.Config, shouldBeAdvertised bool, logger *zap.SugaredLogger) (err error) {
+	logger.Debugf("checking advertisement for service '%s'", serviceName)
+	// Get all config Secrets for this ProxyGroup.
+	// Get all Pods
+	secrets := &corev1.SecretList{}
+	if err := a.List(ctx, secrets, client.InNamespace(a.tsNamespace), client.MatchingLabels(pgSecretLabels(pgName, "config"))); err != nil {
+		return fmt.Errorf("failed to list config Secrets: %w", err)
+	}
+
+	if svc != nil && shouldBeAdvertised {
+		shouldBeAdvertised, err = a.checkEndpointsReady(ctx, svc, logger)
+		if err != nil {
+			return fmt.Errorf("failed to check readiness of Service '%s' endpoints: %w", svc.Name, err)
+		}
+	}
+
+	for _, secret := range secrets.Items {
+		var updated bool
+		for fileName, confB := range secret.Data {
+			var conf ipn.ConfigVAlpha
+			if err := json.Unmarshal(confB, &conf); err != nil {
+				return fmt.Errorf("error unmarshalling ProxyGroup config: %w", err)
+			}
+
+			idx := slices.Index(conf.AdvertiseServices, serviceName.String())
+			isAdvertised := idx >= 0
+			switch {
+			case !isAdvertised && !shouldBeAdvertised:
+				logger.Debugf("service %q shouldn't be advertised", serviceName)
+				continue
+			case isAdvertised && shouldBeAdvertised:
+				logger.Debugf("service %q is already advertised", serviceName)
+				continue
+			case isAdvertised && !shouldBeAdvertised:
+				logger.Debugf("deleting advertisement for service %q", serviceName)
+				conf.AdvertiseServices = slices.Delete(conf.AdvertiseServices, idx, idx+1)
+			case shouldBeAdvertised:
+				replicaName, ok := strings.CutSuffix(secret.Name, "-config")
+				if !ok {
+					logger.Infof("[unexpected] unable to determine replica name from config secret name %q, unable to determine if backend routing has been configured", secret.Name)
+					return nil
+				}
+				ready, err := a.backendRoutesSetup(ctx, serviceName.String(), replicaName, pgName, cfg, logger)
+				if err != nil {
+					return fmt.Errorf("error checking backend routes: %w", err)
+				}
+				if !ready {
+					logger.Debugf("service %q is not ready to be advertised", serviceName)
+					continue
+				}
+				logger.Debugf("advertising service '%s' in secret with name '%s'", serviceName, secret.DeepCopy().Name)
+				conf.AdvertiseServices = append(conf.AdvertiseServices, serviceName.String())
+			}
+			confB, err := json.Marshal(conf)
+			if err != nil {
+				return fmt.Errorf("error marshalling ProxyGroup config: %w", err)
+			}
+			mak.Set(&secret.Data, fileName, confB)
+			updated = true
+		}
+		if updated {
+			logger.Debugf("updating Secret %q", secret.Name)
```

Comment:
This also seems like a redundant line.

We log a bit too much in the operator already. I think we should log just once where it's clear that the state has changed unadvertised -> advertised or vice versa